diff --git a/src/de/fhhannover/inform/iron/mapserver/provider/BasicAuthProviderPropImpl.java b/src/de/fhhannover/inform/iron/mapserver/provider/BasicAuthProviderPropImpl.java
index 510baed..bbfe36f 100644
--- a/src/de/fhhannover/inform/iron/mapserver/provider/BasicAuthProviderPropImpl.java
+++ b/src/de/fhhannover/inform/iron/mapserver/provider/BasicAuthProviderPropImpl.java
@@ -40,7 +40,8 @@ import de.fhhannover.inform.iron.mapserver.utils.NullCheck;
  */
 public class BasicAuthProviderPropImpl implements BasicAuthProvider {
 
-	private PropertiesReaderWriter mProperties;
+	private String mPropertiesFileName;
+
 	
 	public BasicAuthProviderPropImpl(ServerConfigurationProvider serverConf)
 										throws ProviderInitializationException{
@@ -49,8 +50,14 @@ public class BasicAuthProviderPropImpl implements BasicAuthProvider {
 	
 	@Override
 	public boolean verify(String username, String password) {
-		String passFile = mProperties.getProperty(username);
-		return password.equals(passFile);
+		try {
+			PropertiesReaderWriter props = new PropertiesReaderWriter(mPropertiesFileName, true);
+			String passFile = props.getProperty(username);
+			return password.equals(passFile);
+		} catch (IOException e) {
+			return false;
+		}
+
 	}
 
 	private void loadPropertiesFile(ServerConfigurationProvider serverConf)
@@ -62,7 +69,8 @@ public class BasicAuthProviderPropImpl implements BasicAuthProvider {
 				throw new ProviderInitializationException(
 						"basic auth user/pass file not given");
 			}
-			mProperties = new PropertiesReaderWriter(fileName, true);
+			PropertiesReaderWriter testProps = new PropertiesReaderWriter(fileName, true);
+                        mPropertiesFileName = fileName;
 		} catch (IOException e) {
 			throw new ProviderInitializationException(e.getMessage());
 		}
