diff --git a/src/de/fhhannover/inform/iron/mapserver/binding/JaxbRequestUnmarshaller.java b/src/de/fhhannover/inform/iron/mapserver/binding/JaxbRequestUnmarshaller.java
index 510baed..bbfe36f 100644
--- a/src/de/fhhannover/inform/iron/mapserver/binding/JaxbRequestUnmarshaller.java
+++ b/src/de/fhhannover/inform/iron/mapserver/binding/JaxbRequestUnmarshaller.java
@@ -23,6 +23,7 @@ package de.fhhannover.inform.iron.mapserver.binding;
 
 import java.io.IOException;
 import java.io.InputStream;
+import java.io.InputStreamReader;
 import java.math.BigInteger;
 import java.util.ArrayList;
 import java.util.HashMap;
@@ -60,6 +61,7 @@ import org.w3c.dom.Element;
 import org.w3c.dom.NamedNodeMap;
 import org.w3c.dom.Node;
 import org.xml.sax.SAXException;
+import org.xml.sax.InputSource;
 
 import de.fhhannover.inform.iron.mapserver.datamodel.identifiers.Identifier;
 import de.fhhannover.inform.iron.mapserver.datamodel.meta.Metadata;
@@ -171,7 +173,17 @@ class JaxbRequestUnmarshaller implements RequestUnmarshaller {
 		Envelope reqEnv = null;
 		Body soapBody = null;
 		Document requestDocument = null;
-		
+                InputSource is = null;
+
+                try {
+
+                        InputStreamReader inputReader = new InputStreamReader(requestContent,"UTF-8");
+                        is = new InputSource(inputReader);
+                        is.setEncoding("UTF-8");
+                } catch (IOException e) {
+                        e.printStackTrace();
+                }
+
 		mBinder.set(mJaxbCtx.createBinder());
 	
 		// Do we want to validate against the given schema?
@@ -186,7 +198,7 @@ class JaxbRequestUnmarshaller implements RequestUnmarshaller {
 		}
 		
 		try {
-			requestDocument = mDocumentBuilder.get().parse(requestContent);
+			requestDocument = mDocumentBuilder.get().parse(is);
 			reqEnvJaxb = (JAXBElement<Envelope>) mBinder.get().unmarshal(requestDocument);
 		} catch (SAXException e) {
 			e.printStackTrace();
